app/models/uphold_connection.rb

```diff
@@ -0,0 +1,130 @@
+  end
+
+  belongs_to :publisher
+
+  # uphold_code is an intermediate step to acquiring uphold_access_parameters
+  # and should be cleared once it has been used to get uphold_access_parameters
+  validates :uphold_code, absence: true, if: -> { uphold_access_parameters.present? || uphold_verified? }
+
+  # uphold_access_parameters should be cleared once uphold_verified has been set
+  # (see `verify_uphold` method below)
+  validates :uphold_access_parameters, absence: true, if: -> { uphold_verified? }
+
+  # publishers that have uphold codes that have been sitting for five minutes
+  # can be cleared if publishers do not create wallet within 5 minute window
+  scope :has_stale_uphold_code, -> {
+    where.not(encrypted_uphold_code: nil).
+      where("updated_at < ?", UPHOLD_CODE_TIMEOUT.ago)
+  }
+
+  # publishers that have access params that havent accepted by eyeshade
+  # can be cleared after 2 hours
+  scope :has_stale_uphold_access_parameters, -> {
+    where.not(encrypted_uphold_access_parameters: nil).
+      where("updated_at < ?", UPHOLD_ACCESS_PARAMS_TIMEOUT.ago)
+  }
+
+  # This state token is generated and must be unique when connecting to uphold.
+  # It is used to navigate to Uphold, therefore on GET request this state token must be there.
+  def prepare_uphold_state_token
+    return if uphold_state_token.present?
+    self.uphold_state_token = SecureRandom.hex(64).to_s
+    save!
+  end
+
+  def receive_uphold_code(code)
+    self.uphold_state_token = nil
+    self.uphold_code = code
+    self.uphold_access_parameters = nil
+    self.uphold_verified = false
+    save!
+  end
+
+  def verify_uphold
+    self.uphold_state_token = nil
+    self.uphold_code = nil
+    self.uphold_access_parameters = nil
+    self.uphold_verified = true
+    save!
+  end
+
+  def disconnect_uphold
+    self.uphold_code = nil
+    self.uphold_access_parameters = nil
+    self.uphold_verified = false
+    save!
+  end
+
+  def uphold_reauthorization_needed?
+    uphold_verified? &&
+      wallet.present? &&
+      ['re-authorize', 'authorize'].include?(wallet.action)
+  end
+
+  def uphold_status
+    if publisher&.wallet&.uphold_account_status&.to_sym == UpholdAccountState::BLOCKED
+      # Notify on Slack that there's someone suspect
+      SlackMessenger.new(message: "Publisher #{id} is blocked by Uphold and has just logged in. <!channel>").perform
+    end
+
+    if publisher&.wallet&.uphold_account_status&.to_sym == UpholdAccountState::RESTRICTED
+      UpholdAccountState::RESTRICTED
+    elsif uphold_verified?
+      if uphold_reauthorization_needed?
+        UpholdAccountState::REAUTHORIZATION_NEEDED
+      elsif publisher&.wallet&.not_a_member?
+        UpholdAccountState::RESTRICTED
+      else
+        UpholdAccountState::VERIFIED
+      end
+    elsif uphold_access_parameters.present?
+      :access_parameters_acquired
+    elsif uphold_code.present?
+      :code_acquired
+    else
+      UpholdAccountState::UNCONNECTED
+    end
+  end
```

Comment:
Maybe this function could use some cleanup , consider something along the lines of this

lets make this a variable 

```ruby 
uphold_account_status = publisher&.wallet&.uphold_account_status&.to_sym
```

then let's put SlackMessenger in a function

```ruby
send_blocked_message if uphold_account_status == UpholdAccountState::BLOCKED
```

then 

```ruby 
if uphold_account_status == UpholdAccountState::RESTRICTED
    UpholdAccountState::RESTRICTED

elsif uphold_verified? && uphold_reauthorization_needed?
    UpholdAccountState::REAUTHORIZATION_NEEDED

elsif uphold_verified? && publisher&.wallet&.not_a_member?
    UpholdAccountState::RESTRICTED

elsif uphold_verified? && !uphold_reauthorization_needed? && !publisher&.wallet&.not_a_member?
    UpholdAccountState::VERIFIED

elsif uphold_access_parameters.present?
    :access_parameters_acquired

elsif uphold_code.present?
    :code_acquired

else
    UpholdAccountState::UNCONNECTED
end

Reply:
Glad you proposed this. I was debating changing it now or waiting till later. It looks really good

https://github.com/brave-intl/publishers/pull/1825/commits/ef4d1491cd420fd18872b426157f70e123dd8d91